site/docs/plugins/menu.md

```diff
@@ -93,57 +93,373 @@ bot.start();
+This will also set `main` as the parent of `settings`.
+The parent menu is used when backwards navigation is performed.
+You can set a different parent, too.
+
+```ts
+// Register settings menu at main menu
+main.register(settings);
+// Optionally, set a different parent
+main.register(settings, "back-from-settings-menu");
+```
 
-TODO: code examples, explain differences, explain lazy navigation, mention immediate mode
+You can register as many menus as you like, and nest them as deeply as you like.
+The menu identifiers let you jump easily to any page.
 
-## Backwards Navigation
+**Note that you only have to make a single menu of your structure interactive.**
+For example, only pass the root menu to `bot.use`.
 
-Use `ctx.menu.back()` or `menuA.back()`.
+```ts
+// Do this:
+bot.use(main);
 
-TODO: code examples
+// Don't do this:
+bot.use(main);
+bot.use(settings);
+```
 
 ## Payloads
 
-You can store short text along with all text and navigation buttons.
-Use them in your handlers from `ctx.match`.
+You can store short texts along with all text and navigation buttons.
+When the respective handlers are invoked, the payload will be available under `ctx.match`.
+This is useful because it lets you store a little bit of data in a menu.
 
-TODO: what problem does this solve, and give code examples
+Here is an example menu that remembers its creator in the payload.
+Other use cases could for example be to store the index in a pagination.
+
+```ts
+const menu = new Menu("pun-intended")
+  .text(
+    { text: "I know my creator", payload: (ctx) => ctx.from.first_name },
+    (ctx) => ctx.reply(`I was once created by ${ctx.match}!`),
+  );
+
+bot.use(menu);
+bot.command("menu", async (ctx) => {
+  await ctx.reply("You summoned a demenu!", { reply_markup: menu });
+});
+```
+
+Payloads also work well together with dynamic ranges.
 
 ## Dynamic Ranges
 
-Use `menuA.dynamic` to add a dynamic number of buttons.
-Use a `MenuRange` for this.
+So far, we've only seen how to change the text on a button dynamically.
+You can also dynamically adjust the structure of a menu in order to add and remove buttons on the fly.
+
+::: danger Never Change a Menu During Message Handling
+You cannot create or change menus your during message handling.
+All menus must be fully created and registered before your bot starts.
+
+Adding new menus while your bot is running would cause a memory leak.
+Your bot would slow down more and more, and eventually crash.
+:::
+
+You can let a part of a menu's buttons be generated on the fly (or all of them if you want).
+We call this part of the menu a _dynamic range_.
+The easiest way to create a dynamic range is by using the `MenuRange` class that this plugin provides.
+A `MenuRange` provides you with exactly the same functions as a menu, but it does not have an identifier, and it cannot be registered.
+
+```ts
+function getRandomInt(min: number, max: number) {
+  return min + Math.floor(Math.random() * (max - min));
```

Comment:
Oh, okay... I guess this is considered as "normal" so will be assumed to work that way?

To me it seems like really unexpected behavior. Maybe we can add a comment to avoid this issue:
```ts
// Returns a random number between and min [inclusive] and max [exclusive] 
```

Reply:
Yes, that's default behaviour. Check python's `range`, and `String.prototype.substring` in JS, and all other things in all languages.
Note also that `Math.random()` itself returns values from `[0, 1)`. We just reflect that. In other words, if the 1 would be inclusive, `max` would automatically be inclusive, too.

I'm okay with adding a comment if that's helpful. I'm only hesitant because random numbers are not the point of the example.

Reply:
How about calling the arguments `minInclusive` and `maxExclusive`?

Reply:
Please have a look at 40a13a6

Reply:
That's perfect and much better than a comment.

I was originally going to suggest a change like this, but thought it would be less likely to be accepted than just adding a comment.